src/mito2/src/memtable/bulk/part.rs

```diff
@@ -0,0 +1,666 @@
+        }
+
+        let part_size = { dest.data.len() };
+        dest.metadata = BulkPartMeta {
+            num_rows: arrow_record_batch.num_rows(),
+            size: part_size,
+            max_timestamp: max_ts,
+            min_timestamp: min_ts,
+        };
+        Ok(())
+    }
+
+    /// Decodes [BulkPart] to [Batch]es.
+    fn decode_to_batches(&self, _part: &BulkPart, _dest: &mut VecDeque<Batch>) -> Result<()> {
+        todo!()
+    }
+}
+
+/// Converts mutations to record batches
+fn mutations_to_record_batch(
+    mutations: &[Mutation],
+    metadata: &RegionMetadataRef,
+    pk_encoder: &McmpRowCodec,
+    dedup: bool,
+) -> Result<(RecordBatch, i64, i64)> {
+    let total_rows: usize = mutations
+        .iter()
+        .map(|m| m.rows.as_ref().map(|r| r.rows.len()).unwrap_or(0))
+        .sum();
+
+    let mut pk_builder = BinaryBuilder::with_capacity(total_rows, 0);
+
+    let mut ts_vector: Box<dyn MutableVector> = metadata
+        .time_index_column()
+        .column_schema
+        .data_type
+        .create_mutable_vector(total_rows);
+    let mut sequence_vector = UInt64Builder::with_capacity(total_rows);
+    let mut op_type_vector = UInt8Builder::with_capacity(total_rows);
+
+    let mut field_builders: Vec<Box<dyn MutableVector>> = metadata
+        .field_columns()
+        .map(|f| f.column_schema.data_type.create_mutable_vector(total_rows))
+        .collect();
+
+    let mut pk_buffer = vec![];
+    for m in mutations {
+        let Some(key_values) = KeyValuesRef::new(metadata, m) else {
+            continue;
+        };
+
+        for row in key_values.iter() {
+            pk_buffer.clear();
+            pk_encoder.encode_to_vec(row.primary_keys(), &mut pk_buffer)?;
+            pk_builder.append_value(pk_buffer.as_bytes());
+            ts_vector.push_value_ref(row.timestamp());
+            sequence_vector.append_value(row.sequence());
+            op_type_vector.append_value(row.op_type() as u8);
+            for (builder, field) in field_builders.iter_mut().zip(row.fields()) {
+                builder.push_value_ref(field);
+            }
+        }
+    }
+
+    let schema = to_sst_arrow_schema(metadata);
+
+    sort_arrays_to_record_batch(
+        metadata,
+        pk_builder.finish(),
+        ts_vector.to_vector().to_arrow_array(),
+        sequence_vector.finish(),
+        op_type_vector.finish(),
+        field_builders
+            .iter_mut()
+            .map(|f| f.to_vector().to_arrow_array()),
+        schema,
+        dedup,
+    )
+}
+
+/// Converts arrays to arrow record batch with SST schema along with timestamp range.
+#[allow(clippy::too_many_arguments)]
+fn sort_arrays_to_record_batch(
+    metadata: &RegionMetadataRef,
+    pk: BinaryArray,
+    timestamp: ArrayRef,
+    sequence: UInt64Array,
+    op_type: UInt8Array,
+    fields: impl Iterator<Item = ArrayRef>,
+    schema: SchemaRef,
+    dedup: bool,
+) -> Result<(RecordBatch, i64, i64)> {
```

Comment:
nitpick: What about refactoring it into a struct? This fixes the clippy warning.
```rust
struct ArraysSorter<'a, I> {
    metadata: &'a RegionMetadataRef,
    fields: I,
}

impl<'a, I: Iterator<Item = ArrayRef>> ArraysSorter<'a, I> {
    fn to_record_batch(&self) -> Result<(RecordBatch, i64, i64)> {}
}
```
What's more, we could only keep the time index data type as a field instead of taking the whole `RegionMetadataRef`.

Reply:
Introduced an `ArraysSorter` here https://github.com/GreptimeTeam/greptimedb/pull/4163/commits/e46d05fb6a67dd865b8f1f02f15e46e82c3ed1cd#diff-176b5ab04e2d4dfe92ae84d90f94a03fe5e9eb2bf721511cd991d47a8408548cR167